test/TestUtils/TestProjectFixture.cs

```diff
@@ -0,0 +1,362 @@
+            _sharedLibraryPrefix = sharedLibraryPrefix 
+                ?? RuntimeInformationExtensions.GetSharedLibraryPrefixForCurrentPlatform();
+
+            _repoDirectoriesProvider = repoDirectoriesProvider;
+
+            _testProjectSourceDirectory = testProjectSourceDirectory
+                ?? Path.Combine(repoDirectoriesProvider.RepoRoot, "TestAssets", "TestProjects");
+            _testArtifactDirectory = _testArtifactDirectory
+                ?? Environment.GetEnvironmentVariable(s_testArtifactDirectoryEnvironmentVariable)
+                ?? Path.Combine(AppContext.BaseDirectory, s_testArtifactDirectoryEnvironmentVariable);
+
+            _sdkDotnet = new DotNetCli(dotnetInstallPath ?? DotNetCli.GetStage0Path(repoDirectoriesProvider.RepoRoot));
+            _currentRid = currentRid ?? _sdkDotnet.GetRuntimeId();
+            
+            _builtDotnet = new DotNetCli(repoDirectoriesProvider.BuiltDotnet);
+
+            InitializeTestProject(
+                _testProjectName,
+                _testProjectSourceDirectory, 
+                _testArtifactDirectory,
+                _exeExtension,
+                _sharedLibraryExtension,
+                _sharedLibraryPrefix);
+        }
+
+        public TestProjectFixture(TestProjectFixture fixtureToCopy)
+        {
+            _testProjectName = fixtureToCopy._testProjectName;
+            _exeExtension = fixtureToCopy._exeExtension;
+            _sharedLibraryExtension = fixtureToCopy._sharedLibraryExtension;
+            _sharedLibraryPrefix = fixtureToCopy._sharedLibraryPrefix;
+            _repoDirectoriesProvider = fixtureToCopy._repoDirectoriesProvider;
+            _testProjectSourceDirectory = fixtureToCopy._testProjectSourceDirectory;
+            _testArtifactDirectory = fixtureToCopy._testArtifactDirectory;
+            _sdkDotnet = fixtureToCopy._sdkDotnet;
+            _currentRid = fixtureToCopy._currentRid;
+            _builtDotnet = fixtureToCopy._builtDotnet;
+            _sourceTestProject = fixtureToCopy._sourceTestProject;
+
+            _testProject = CopyTestProject(
+                fixtureToCopy.TestProject, 
+                _testArtifactDirectory, 
+                _exeExtension,
+                _sharedLibraryExtension,
+                _sharedLibraryPrefix);
+        }
+
+        private void InitializeTestProject(
+            string testProjectName,
+            string testProjectSourceDirectory,
+            string testArtifactDirectory,
+            string exeExtension,
+            string sharedLibraryExtension,
+            string sharedLibraryPrefix)
+        {
+            var sourceTestProjectPath = Path.Combine(testProjectSourceDirectory, testProjectName);
+            _sourceTestProject = new TestProject(
+                sourceTestProjectPath,
+                exeExtension,
+                sharedLibraryExtension,
+                sharedLibraryPrefix);
+
+            _testProject = CopyTestProject(
```

Comment:
> TestTargets.CopyTestProjectsToArtifacts

I don't actually think this is relevant anymore, will remove it. The first time a TestProjectFixture is created it creates a copy of the one in TestAssets/TestProjects, builds it, and ensures it is restored. This is then copied and mutated for specific scenarios. The first copy may not be necessary but keeps test runs isolated inside the artifacts folder.

---
In reply to: [66347830](https://github.com/dotnet/core-setup/pull/97#discussion_r66347830) [](ancestors = 66347830)